pkg/ipam-node/handlers/allocate.go

```diff
@@ -191,38 +198,79 @@ func (h *Handlers) allocateInPool(poolName string, reqLog logr.Logger,
-			return PoolAlloc{}, status.Errorf(codes.AlreadyExists,
-				"allocation already exist in the pool \"%s\", poolType \"%s\"", poolName, poolType)
-		}
-		if errors.Is(err, allocator.ErrNoFreeAddresses) {
-			return PoolAlloc{}, status.Errorf(codes.ResourceExhausted,
-				"no free addresses in the pool \"%s\", poolType \"%s\"",
-				poolName, poolType)
-		}
-		if errors.Is(err, storePkg.ErrIPAlreadyReserved) {
-			return PoolAlloc{}, status.Errorf(codes.ResourceExhausted,
-				"requested IP is already reserved in the pool \"%s\", poolType \"%s\"",
-				poolName, poolType)
-		}
-		return PoolAlloc{}, status.Errorf(codes.Internal,
-			"failed to allocate IP address in pool \"%s\", poolType \"%s\"", poolName, poolType)
+		return PoolAlloc{}, specificError(err, poolName, poolType)
 	}
 	if params.Features != nil && params.Features.AllocateDefaultGateway {
 		// TODO (ykulazhenkov): do we want to keep gateway in this case?
 		// if we will return gateway here, the container will have same IP as interface address and as gateway
 		result.Gateway = nil
 	}
-
+	routes, err := getRoutes(result.Gateway, poolCfg.Routes, poolCfg.DefaultGateway)
+	if err != nil {
+		return PoolAlloc{}, err
+	}
 	poolLog.Info("IP address allocated", "allocation", result.String())
 	return PoolAlloc{
 		Pool:     poolName,
 		IPConfig: result,
+		Routes:   routes,
 	}, nil
 }
 
+func specificError(err error, poolName string, poolType string) error {
+	if errors.Is(err, storePkg.ErrReservationAlreadyExist) {
+		return status.Errorf(codes.AlreadyExists,
+			"allocation already exist in the pool \"%s\", poolType \"%s\"", poolName, poolType)
+	}
+	if errors.Is(err, allocator.ErrNoFreeAddresses) {
+		return status.Errorf(codes.ResourceExhausted,
+			"no free addresses in the pool \"%s\", poolType \"%s\"",
+			poolName, poolType)
+	}
+	if errors.Is(err, storePkg.ErrIPAlreadyReserved) {
+		return status.Errorf(codes.ResourceExhausted,
+			"requested IP is already reserved in the pool \"%s\", poolType \"%s\"",
+			poolName, poolType)
+	}
+	return status.Errorf(codes.Internal,
+		"failed to allocate IP address in pool \"%s\", poolType \"%s\"", poolName, poolType)
+}
+
 func poolCfgError(reqLog logr.Logger, pool, poolType, reason string) error {
 	reqLog.Error(nil, "invalid pool config", "pool", pool, "poolType", poolType,
 		"reason", reason)
 	return status.Errorf(codes.Internal, "invalid config for pool \"%s\", poolType \"%s\"", pool, poolType)
 }
+
+func getRoutes(gateway net.IP, routesCfg []pool.Route, defaultGateway bool) ([]net.IPNet, error) {
+	routes := make([]net.IPNet, 0)
+	if gateway != nil {
+		for _, r := range routesCfg {
+			_, ipNet, err := net.ParseCIDR(r.Dst)
+			if err != nil {
+				return nil, status.Errorf(codes.InvalidArgument,
+					"unexpected Route destination format, received value: %s, error: %s",
+					r.Dst, err.Error())
+			}
+			routes = append(routes, *ipNet)
+		}
+		if defaultGateway {
+			routes = append(routes, createDefaultRoute(gateway))
+		}
```

Comment:
Maybe sort and compact the slice here to ensure that unique routes are added.

Reply:
Done

Reply:
nit: I think you could use those 2:
* https://pkg.go.dev/slices#Sort
* https://pkg.go.dev/slices#Compact

Reply:
Or https://pkg.go.dev/slices#SortFunc

Reply:
Done